PiT 24.7: in GH windows runner flow fails when using node installed in the prepare-frontend step

### Description of the bug

When running PiT starters in GH actions for windows, they fail when the node installation needs to be upgraded by flow.
Flow `prepare-frontend` installs a valid `node.exe` in `~/.vaadin/node/` but then, it fails when it's used in flow.

Error liness than can be seen in output are:

- `prepare-frontend`
```
[INFO] The globally installed Node.js version 18.20.5 is older than the required minimum version 20.0.0. Using Node.js from C:\Users\runneradmin\.vaadin.
[INFO] Couldn't find node.exe. Installing Node and npm to C:\Users\runneradmin\.vaadin.
[INFO] Installing node version v22.12.0
[INFO] Downloading https://nodejs.org/dist/v22.12.0/node-v22.12.0-win-x64.zip to C:\Users\runneradmin\.vaadin\node-v22.12.0-win-x64.zip
[INFO] No proxies configured. If you are behind a proxy server, see https://vaadin.com/docs/latest/configuration/development-mode/node-js#proxy-settings-for-downloading-frontend-toolchain for information on proxy configuration.
[INFO] Unpacking C:\Users\runneradmin\.vaadin\node-v22.12.0-win-x64.zip (34872043 bytes) into C:\Users\runneradmin\.vaadin\node\tmp
[INFO] Copying node binary from C:\Users\runneradmin\.vaadin\node\tmp\node-v22.12.0-win-x64\node.exe to C:\Users\runneradmin\.vaadin\node\node.exe
[INFO] Extracting npm
[INFO] Local node installation successful.
```

- `mvn spring-boot:run`

```
2024-12-30T11:04:39.592Z ERROR 5792 --- [  restartedMain] c.v.f.s.frontend.FrontendToolsLocator    : Failed to read the command '[where, node.exe]' stdout/stderr

...

2024-12-30T11:04:49.667Z ERROR 5792 --- [  restartedMain] c.v.f.s.frontend.FrontendToolsLocator    : Failed to read the command '[C:\Users\runneradmin\.vaadin\node\node.exe, -v]' stdout/stderr

...
Caused by: java.lang.IllegalStateException: Failed to install Node
	at com.vaadin.flow.server.frontend.FrontendTools.installNode(FrontendTools.java:691) ~[flow-server-24.7.0.alpha2.jar:24.7.0.alpha2]
	at com.vaadin.flow.server.frontend.FrontendTools.getNodeExecutable(FrontendTools.java:439) ~[flow-server-24.7.0.alpha2.jar:24.7.0.alpha2]
	at com.vaadin.exampledata.NodeUtil.runScript(NodeUtil.java:36) ~[exampledata-6.2.0.jar:6.2.0]
	at com.vaadin.exampledata.NodeUtil.initialize(NodeUtil.java:29) ~[exampledata-6.2.0.jar:6.2.0]
	at com.vaadin.exampledata.NodeScriptInterface.init(NodeScriptInterface.java:28) ~[exampledata-6.2.0.jar:6.2.0]
	at com.vaadin.exampledata.NodeScriptInterface.getChanceString(NodeScriptInterface.java:34) ~[exampledata-6.2.0.jar:6.2.0]
	at com.vaadin.exampledata.ChanceStringType.getValue(ChanceStringType.java:22) ~[exampledata-6.2.0.jar:6.2.0]
	at com.vaadin.exampledata.ChanceStringType.getValue(ChanceStringType.java:6) ~[exampledata-6.2.0.jar:6.2.0]
	at com.vaadin.exampledata.ExampleDataGenerator.assignValue(ExampleDataGenerator.java:48) ~[exampledata-6.2.0.jar:6.2.0]
	at com.vaadin.exampledata.ExampleDataGenerator.createBean(ExampleDataGenerator.java:38) ~[exampledata-6.2.0.jar:6.2.0]
	... 25 common frames omitted
Caused by: com.vaadin.flow.server.frontend.installer.InstallationException: Unable to detect version of Node. Using command C:\Users\runneradmin\.vaadin\node\node.exe --version
	at com.vaadin.flow.server.frontend.installer.NodeInstaller.getVersion(NodeInstaller.java:574) ~[flow-server-24.7.0.alpha2.jar:24.7.0.alpha2]
	at com.vaadin.flow.server.frontend.installer.NodeInstaller.nodeIsAlreadyInstalled(NodeInstaller.java:238) ~[flow-server-24.7.0.alpha2.jar:24.7.0.alpha2]
	at com.vaadin.flow.server.frontend.installer.NodeInstaller.install(NodeInstaller.java:214) ~[flow-server-24.7.0.alpha2.jar:24.7.0.alpha2]
	at com.vaadin.flow.server.frontend.FrontendTools.installNode(FrontendTools.java:689) ~[flow-server-24.7.0.alpha2.jar:24.7.0.alpha2]
	... 34 common frames omitted
```

### Expected behavior

It installs and uses correctly node, like it happens in linux/mac

### Minimal reproducible example

In a github actions workflow select `windows-latest` runner and execute the followin command sequence

```
# Clean stuff
rm -rf ~/.m2/repository/
rm -rf ~/.vaadin/node/
# Clone repo
git clone https://github.com/vaadin/designer-tutorial.git
cd designer-tutorial
# Change version
perl -pi -e 's|(\s*<'vaadin.version'>)[^\s]+(</'vaadin.version'>)|${1}24.7.0.alpha1${2}|g' pom.xml
# Run build
mvn -ntp -B
```


### Versions

- Vaadin / Flow version: 24.7.0.alpha1
- OS version: windows-latest in gihub actions

--- a/flow-test-generic/src/main/java/com/vaadin/flow/testutil/ClassesSerializableTest.java
+++ b/flow-test-generic/src/main/java/com/vaadin/flow/testutil/ClassesSerializableTest.java
@@ -79,6 +79,7 @@ public abstract class ClassesSerializableTest extends ClassFinder {
                 "com\\.vaadin\\.base\\.devserver\\.DebugWindowConnection",
                 "com\\.vaadin\\.base\\.devserver\\.DebugWindowConnection\\$DevToolsInterfaceImpl",
                 "com\\.vaadin\\.base\\.devserver\\.DevModeHandlerManagerImpl",
+                "com\\.vaadin\\.base\\.devserver\\.DevModeHandlerManagerImpl\\$InternalThreadFactory",
                 "com\\.vaadin\\.base\\.devserver\\.DevServerWatchDog",
                 "com\\.vaadin\\.base\\.devserver\\.DevServerWatchDog\\$WatchDogServer",
                 "com\\.vaadin\\.base\\.devserver\\.DevToolsInterface",

--- a/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/DevModeHandlerManagerImpl.java
+++ b/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/DevModeHandlerManagerImpl.java
@@ -15,18 +15,20 @@
  */
 package com.vaadin.base.devserver;
 
-import com.vaadin.flow.server.Command;
 import jakarta.servlet.annotation.HandlesTypes;
 
 import java.io.Closeable;
 import java.io.File;
-import java.io.IOException;
 import java.io.Serializable;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicInteger;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -35,6 +37,7 @@ import com.vaadin.base.devserver.startup.DevModeInitializer;
 import com.vaadin.base.devserver.startup.DevModeStartupListener;
 import com.vaadin.flow.internal.DevModeHandler;
 import com.vaadin.flow.internal.DevModeHandlerManager;
+import com.vaadin.flow.server.Command;
 import com.vaadin.flow.server.Mode;
 import com.vaadin.flow.server.VaadinContext;
 import com.vaadin.flow.server.frontend.FrontendUtils;
@@ -68,6 +71,7 @@ public class DevModeHandlerManagerImpl implements DevModeHandlerManager {
     private DevModeHandler devModeHandler;
     private BrowserLauncher browserLauncher;
     private final Set<Command> shutdownCommands = new HashSet<>();
+    private ExecutorService executorService;
 
     private String applicationUrl;
     private boolean fullyStarted = false;
@@ -96,8 +100,11 @@ public class DevModeHandlerManagerImpl implements DevModeHandlerManager {
     @Override
     public void initDevModeHandler(Set<Class<?>> classes, VaadinContext context)
             throws VaadinInitializerException {
-        setDevModeHandler(
-                DevModeInitializer.initDevModeHandler(classes, context));
+        shutdownExecutorService();
+        executorService = Executors.newFixedThreadPool(4,
+                new InternalThreadFactory());
+        setDevModeHandler(DevModeInitializer.initDevModeHandler(classes,
+                context, executorService));
         CompletableFuture.runAsync(() -> {
             DevModeHandler devModeHandler = getDevModeHandler();
             if (devModeHandler instanceof AbstractDevServerRunner) {
@@ -111,9 +118,15 @@ public class DevModeHandlerManagerImpl implements DevModeHandlerManager {
             startWatchingThemeFolder(context, config);
             watchExternalDependencies(context, config);
             setFullyStarted(true);
-        });
+        }, executorService);
         setDevModeStarted(context);
         this.browserLauncher = new BrowserLauncher(context);
+    }
+
+    private void shutdownExecutorService() {
+        if (executorService != null) {
+            executorService.shutdownNow();
+        }
     }
 
     private void watchExternalDependencies(VaadinContext context,
@@ -159,6 +172,7 @@ public class DevModeHandlerManagerImpl implements DevModeHandlerManager {
             devModeHandler.stop();
             devModeHandler = null;
         }
+        shutdownExecutorService();
         for (Command shutdownCommand : shutdownCommands) {
             try {
                 shutdownCommand.execute();
@@ -230,5 +244,19 @@ public class DevModeHandlerManagerImpl implements DevModeHandlerManager {
 
     private static Logger getLogger() {
         return LoggerFactory.getLogger(DevModeHandlerManagerImpl.class);
+    }
+
+    private static class InternalThreadFactory implements ThreadFactory {
+        private final AtomicInteger threadNumber = new AtomicInteger(1);
+
+        @Override
+        public Thread newThread(Runnable runnable) {
+            String threadName = "vaadin-dev-server-"
+                    + threadNumber.getAndIncrement();
+            Thread thread = new Thread(runnable, threadName);
+            thread.setDaemon(true);
+            thread.setPriority(Thread.NORM_PRIORITY);
+            return thread;
+        }
     }
 }

--- a/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/DevModeHandlerManagerImpl.java
+++ b/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/DevModeHandlerManagerImpl.java
@@ -126,6 +126,7 @@ public class DevModeHandlerManagerImpl implements DevModeHandlerManager {
     private void shutdownExecutorService() {
         if (executorService != null) {
             executorService.shutdownNow();
+            executorService = null;
         }
     }
 

--- a/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/startup/DevModeInitializer.java
+++ b/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/startup/DevModeInitializer.java
@@ -39,6 +39,8 @@ import java.util.List;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
@@ -73,9 +75,6 @@ import com.vaadin.flow.server.frontend.scanner.ClassFinder.DefaultClassFinder;
 import com.vaadin.flow.server.startup.ApplicationConfiguration;
 import com.vaadin.flow.server.startup.VaadinInitializerException;
 import com.vaadin.pro.licensechecker.LicenseChecker;
-
-import elemental.json.Json;
-import elemental.json.JsonObject;
 
 import static com.vaadin.flow.server.Constants.PACKAGE_JSON;
 import static com.vaadin.flow.server.Constants.PROJECT_FRONTEND_GENERATED_DIR_TOKEN;
@@ -169,6 +168,12 @@ public class DevModeInitializer implements Serializable {
     /**
      * Initialize the devmode server if not in production mode or compatibility
      * mode.
+     * <p>
+     * </p>
+     * Uses common ForkJoin pool to execute asynchronous tasks. It is
+     * recommended to use
+     * {@link #initDevModeHandler(Set, VaadinContext, Executor)} and provide a a
+     * custom executor if initialization starts long-running tasks.
      *
      * @param classes
      *            classes to check for npm- and js modules
@@ -179,9 +184,34 @@ public class DevModeInitializer implements Serializable {
      *
      * @throws VaadinInitializerException
      *             if dev mode can't be initialized
+     * @deprecated use {@link #initDevModeHandler(Set, VaadinContext, Executor)}
+     *             providing a custom executor.
      */
+    @Deprecated(forRemoval = true)
     public static DevModeHandler initDevModeHandler(Set<Class<?>> classes,
             VaadinContext context) throws VaadinInitializerException {
+        return initDevModeHandler(classes, context, ForkJoinPool.commonPool());
+    }
+
+    /**
+     * Initialize the devmode server if not in production mode or compatibility
+     * mode.
+     *
+     * @param classes
+     *            classes to check for npm- and js modules
+     * @param context
+     *            VaadinContext we are running in
+     * @param taskExecutor
+     *            the executor to use for asynchronous execution
+     * @return the initialized dev mode handler or {@code null} if none was
+     *         created
+     *
+     * @throws VaadinInitializerException
+     *             if dev mode can't be initialized
+     */
+    public static DevModeHandler initDevModeHandler(Set<Class<?>> classes,
+            VaadinContext context, Executor taskExecutor)
+            throws VaadinInitializerException {
 
         ApplicationConfiguration config = ApplicationConfiguration.get(context);
         if (config.isProductionMode()) {
@@ -317,7 +347,7 @@ public class DevModeInitializer implements Serializable {
         };
 
         CompletableFuture<Void> nodeTasksFuture = CompletableFuture
-                .runAsync(runnable);
+                .runAsync(runnable, taskExecutor);
 
         Lookup devServerLookup = Lookup.compose(lookup,
                 Lookup.of(config, ApplicationConfiguration.class));

